keylime/tpm2.py

```diff
@@ -568,11 +599,12 @@ def __create_aik(self, activate, asym_alg=None, hash_alg=None, sign_alg=None):
                 'opw': owner_pw,
                 'apw': aik_pw
             }
-            if legacy_tools:
+
+            if tools_version == "3.2":
                 command = "tpm2_getpubak -E {ekhandle} -k 0x81010008 -g {asymalg} -D {hashalg} -s {signalg} -f {akpubfile} -e {epw} -P {apw} -o {opw}".format(**cmdargs)
-            else:
-                command = "tpm2_createak -C {ekhandle} -k - -G {asymalg} -D {hashalg} -s {signalg} -p {akpubfile} -f pem -e {epw} -P {apw} -o {opw}".format(**cmdargs)
-            retDict = self.__run(command, outputpaths=akpubfile.name)
+            elif tools_version == "4.0":
+                command = "tpm2_createak -Q -C {ekhandle} -c ak.ctx -G {asymalg} -g {hashalg} -s {signalg} -u {akpubfile} -f pem -n ak.name -p {apw} -P {epw}".format(**cmdargs)
+            retDict = self.__run(command)
```

Comment:
I think `, outputpaths=akpubfile.name` needs to stay, doesn't it?  We also need to add one for `ak.ctx`.

Reply:
The issue now is that a `handle` is not returned when using `tpm2_createak`, unlike `tpm2_getpubak`.


https://github.com/tpm2-software/tpm2-tools/blob/3.0.X/man/tpm2_getpubak.1.md :

```
loaded-key:
  handle: 800000ff
  name: 000bac149518baa05540a0678bd9b624f8a98d042e46c60f4d098ba394d36fc49268
```

https://github.com/tpm2-software/tpm2-tools/blob/4.0/man/tpm2_createak.1.md :

```
loaded-key:
  name: 000bac149518baa05540a0678bd9b624f8a98d042e46c60f4d098ba394d36fc49268
```

This then throws out the following conditional:

https://github.com/keylime/keylime/blob/b133f9ee95abc4a8dbdf22542069152c2206148f/keylime/tpm2.py#L625-L626

Here is the result against 4.0 tools (with `retout` as debug):

```
Using config file /etc/keylime.conf
2019-10-14 09:35:15.122 - keylime.tpm2 - INFO - TPM2-TOOLS Version: 4.0
2019-10-14 09:35:15.274 - keylime.tpm - WARNING - INSECURE: Keylime is using a software TPM emulator rather than a real hardware TPM.
2019-10-14 09:35:15.274 - keylime.tpm - WARNING - INSECURE: The security of Keylime is NOT linked to a hardware root of trust.
2019-10-14 09:35:15.274 - keylime.tpm - WARNING - INSECURE: Only use Keylime in this mode for testing or debugging purposes.
2019-10-14 09:35:15.275 - keylime.tpm2 - INFO - Taking ownership with config provided TPM owner password: keylime
2019-10-14 09:35:15.573 - keylime.tpm2 - INFO - TPM Owner password confirmed: keylime
2019-10-14 09:35:15.967 - keylime.tpm2 - WARNING - No EK certificate found in TPM NVRAM
retout: [b'loaded-key:\n', b'  name: 000bcc2e02cb58649158ac32d122d8499748982b6ae84ae7ea14b28bd59919a75a51\n']
Traceback (most recent call last):
  File "/usr/local/bin/keylime_agent", line 11, in <module>
    load_entry_point('keylime==1.2', 'console_scripts', 'keylime_agent')()
  File "/usr/local/lib/python3.8/site-packages/keylime-1.2-py3.8.egg/keylime/cloud_agent.py", line 491, in main
    (ek,ekcert,aik,ek_tpm,aik_name) = tpm.tpm_init(self_activate=False,config_pw=config.get('cloud_agent','tpm_ownerpassword')) # this tells initialize not to self activate the AIK
  File "/usr/local/lib/python3.8/site-packages/keylime-1.2-py3.8.egg/keylime/tpm2.py", line 861, in tpm_init
    self.__create_aik(self_activate)
  File "/usr/local/lib/python3.8/site-packages/keylime-1.2-py3.8.egg/keylime/tpm2.py", line 626, in __create_aik
    raise Exception("tpm2_createak failed to create aik: return "+str(retout))
Exception: tpm2_createak failed to create aik: return [b'loaded-key:\n', b'  name: 000bcc2e02cb58649158ac32d122d8499748982b6ae84ae7ea14b28bd59919a75a51\n']
```

And 3.2:

```
Using config file /etc/keylime.conf
2019-10-14 09:35:20.605 - keylime.tpm2 - INFO - TPM2-TOOLS Version: 3.2.0
2019-10-14 09:35:20.673 - keylime.tpm - WARNING - INSECURE: Keylime is using a software TPM emulator rather than a real hardware TPM.
2019-10-14 09:35:20.673 - keylime.tpm - WARNING - INSECURE: The security of Keylime is NOT linked to a hardware root of trust.
2019-10-14 09:35:20.673 - keylime.tpm - WARNING - INSECURE: Only use Keylime in this mode for testing or debugging purposes.
2019-10-14 09:35:20.673 - keylime.tpm2 - INFO - Taking ownership with config provided TPM owner password: keylime
2019-10-14 09:35:20.754 - keylime.tpm2 - INFO - TPM Owner password confirmed: keylime
2019-10-14 09:35:20.957 - keylime.tpm2 - WARNING - No EK certificate found in TPM NVRAM
retout: [b'loaded-key:\n', b'  handle: 800000ff\n', b'  name: 000b1afb8d67fad9349d18ff3ddd03afd7d687ac70903b534c6298fb417e1af66111\n']
2019-10-14 09:35:21.371 - keylime.cloudagent - INFO - Agent UUID: D432FBB3-D2F1-4A97-9EF7-75BD81C00000
2019-10-14 09:35:21.459 - keylime.registrar_client - INFO - Agent registration requested for D432FBB3-D2F1-4A97-9EF7-75BD81C00000
2019-10-14 09:35:21.512 - keylime.tpm2 - INFO - AIK activated.
2019-10-14 09:35:21.520 - keylime.registrar_client - INFO - Registration activated for agent D432FBB3-D2F1-4A97-9EF7-75BD81C00000.
2019-10-14 09:35:21.556 - keylime.cloudagent - INFO - Starting Cloud Agent on port 9002 use <Ctrl-C> to stop
2019-10-14 09:35:21.557 - keylime.revocation_notifier - INFO - Waiting for revocation messages on 0mq 127.0.0.1:8992
```